File: packages/totem-sdk/packages/kissvm/src/vm.ts
import { KissvmLimitError } from './errors.js';
import type { Value, Scalar, ScriptWitness, TxContext, ASTNode } from './types.js';

export const LIMITS = {
  MAX_INSTRUCTIONS: 1024,
  MAX_STACK_DEPTH: 64,
  MAX_PARAMS: 32,
  MAX_STRING_BYTES: 65536,
  MAX_SHIFT_BITS: 256,
} as const;

export interface UserFunction {
  params: string[];
  body: ASTNode[];
  closureEnv: Map<string, Value>;
}

export interface MastBranch {
  hash: string;
  body: ASTNode[];
}

export class VMState {
  /** Instruction counter shared across all frames (including recursive MAST) */
  instructionCount = 0;

  /**
   * Variable environment stack.
   * KISSVM uses flat (script-level) scoping: LET updates the nearest enclosing
   * binding rather than shadowing it.  A NEW frame is only pushed for user-
   * defined function calls (pushCallFrame).  Block constructs (IF/FOR/SWITCH)
   * use pushBlockScope which only increments the depth counter without creating
   * a new env frame.
   */
  private envStack: Map<string, Value>[] = [new Map()];

  /** Call-stack depth guard */
  private callDepth = 0;

  /** User-defined functions */
  readonly funcs: Map<string, UserFunction> = new Map();

  /** Execution trace lines */
  readonly trace: string[] = [];

  /**
   * Branches stored by the most recent MAST { … } definition block.
   * Used by EXEC MAST to select and run a branch.
   */
  mastBlock?: MastBranch[];

  constructor(
    public readonly witness: ScriptWitness,
    public readonly txCtx: TxContext,
  ) {}

  // ─── Scope helpers ────────────────────────────────────────────────────────

  get(name: string): Value | undefined {
    for (let i = this.envStack.length - 1; i >= 0; i--) {
      const v = this.envStack[i].get(name);
      if (v !== undefined) return v;
    }
    return undefined;
  }

  /**
   * LET semantics: if the variable already exists in ANY enclosing scope,
   * update it there (flat/upward assignment).  Otherwise create in current scope.
   */
  set(name: string, value: Value): void {
    for (let i = this.envStack.length - 1; i >= 0; i--) {
      if (this.envStack[i].has(name)) {
        this.envStack[i].set(name, value);
        return;
      }
    }
    this.envStack[this.envStack.length - 1].set(name, value);
  }

  /** Push a new scope for a function call */
  pushCallFrame(): void {
    this.callDepth++;
    if (this.callDepth > LIMITS.MAX_STACK_DEPTH) {
      throw new KissvmLimitError(`Call stack depth exceeded (max ${LIMITS.MAX_STACK_DEPTH})`);
    }
    this.envStack.push(new Map());
  }

  popCallFrame(): void {
    this.callDepth = Math.max(0, this.callDepth - 1);
    if (this.envStack.length > 1) this.envStack.pop();
  }

  /**
   * Block-level scope used only to enforce the depth limit for deeply nested
   * IF/FOR/SWITCH blocks.  Does NOT push a new env frame (flat scoping).
   */
  pushBlockScope(): void {
    this.callDepth++;
    if (this.callDepth > LIMITS.MAX_STACK_DEPTH) {
      throw new KissvmLimitError(`Stack depth exceeded (max ${LIMITS.MAX_STACK_DEPTH})`);
    }
  }

  popBlockScope(): void {
    this.callDepth = Math.max(0, this.callDepth - 1);
  }

  // ─── Instruction counting ─────────────────────────────────────────────────

  tick(n = 1): void {
    this.instructionCount += n;
    if (this.instructionCount > LIMITS.MAX_INSTRUCTIONS) {
      throw new KissvmLimitError(`Instruction limit exceeded (max ${LIMITS.MAX_INSTRUCTIONS})`);
    }
  }

  addTrace(msg: string): void {
    this.trace.push(msg);
  }

  // ─── Snapshot current scope for closures ─────────────────────────────────

  snapshotEnv(): Map<string, Value> {
    const snap = new Map<string, Value>();
    for (const frame of this.envStack) {
      for (const [k, v] of frame) snap.set(k, v);
    }
    return snap;
  }
}

File: packages/totem-sdk/packages/kissvm/rust/src/vm.rs
use std::collections::HashMap;
use crate::types::*;

pub const MAX_INSTRUCTIONS: u32 = 1024;
pub const MAX_STACK_DEPTH: u32 = 64;
pub const MAX_PARAMS: usize = 32;
pub const MAX_STRING_BYTES: usize = 65536;
pub const MAX_SHIFT_BITS: u32 = 256;

#[derive(Debug, Clone)]
pub struct UserFunction {
    pub params: Vec<String>,
    pub body: Vec<AstNode>,
    pub closure_env: HashMap<String, VmValue>,
}

#[derive(Debug, Clone)]
pub struct MastBranch {
    pub hash: String,
    pub body: Vec<AstNode>,
}

pub struct VmState {
    pub instruction_count: u32,
    env_stack: Vec<HashMap<String, VmValue>>,
    call_depth: u32,
    pub funcs: HashMap<String, UserFunction>,
    pub trace: Vec<String>,
    pub mast_block: Option<Vec<MastBranch>>,
    pub witness: ScriptWitness,
    pub tx_ctx: TxContext,
    pub sha3_256: fn(&[u8]) -> Vec<u8>,
    pub sha256: fn(&[u8]) -> Vec<u8>,
    pub wots_verify: fn(&[u8], &[u8], &[u8]) -> bool,
}

impl VmState {
    pub fn new(
        witness: ScriptWitness,
        tx_ctx: TxContext,
        sha3_256: fn(&[u8]) -> Vec<u8>,
        sha256: fn(&[u8]) -> Vec<u8>,
        wots_verify: fn(&[u8], &[u8], &[u8]) -> bool,
    ) -> Self {
        VmState {
            instruction_count: 0,
            env_stack: vec![HashMap::new()],
            call_depth: 0,
            funcs: HashMap::new(),
            trace: Vec::new(),
            mast_block: None,
            witness,
            tx_ctx,
            sha3_256,
            sha256,
            wots_verify,
        }
    }

    pub fn get(&self, name: &str) -> Option<VmValue> {
        for frame in self.env_stack.iter().rev() {
            if let Some(v) = frame.get(name) {
                return Some(v.clone());
            }
        }
        None
    }

    pub fn set(&mut self, name: &str, value: VmValue) {
        for frame in self.env_stack.iter_mut().rev() {
            if frame.contains_key(name) {
                frame.insert(name.to_string(), value);
                return;
            }
        }
        if let Some(last) = self.env_stack.last_mut() {
            last.insert(name.to_string(), value);
        }
    }

    pub fn push_call_frame(&mut self) -> Result<(), String> {
        self.call_depth += 1;
        if self.call_depth > MAX_STACK_DEPTH {
            return Err(format!("Call stack depth exceeded (max {})", MAX_STACK_DEPTH));
        }
        self.env_stack.push(HashMap::new());
        Ok(())
    }

    pub fn pop_call_frame(&mut self) {
        self.call_depth = self.call_depth.saturating_sub(1);
        if self.env_stack.len() > 1 {
            self.env_stack.pop();
        }
    }

    pub fn push_block_scope(&mut self) -> Result<(), String> {
        self.call_depth += 1;
        if self.call_depth > MAX_STACK_DEPTH {
            return Err(format!("Stack depth exceeded (max {})", MAX_STACK_DEPTH));
        }
        Ok(())
    }

    pub fn pop_block_scope(&mut self) {
        self.call_depth = self.call_depth.saturating_sub(1);
    }

    pub fn tick(&mut self, n: u32) -> Result<(), String> {
        self.instruction_count += n;
        if self.instruction_count > MAX_INSTRUCTIONS {
            return Err(format!("Instruction limit exceeded (max {})", MAX_INSTRUCTIONS));
        }
        Ok(())
    }

    pub fn add_trace(&mut self, msg: String) {
        self.trace.push(msg);
    }

    pub fn snapshot_env(&self) -> HashMap<String, VmValue> {
        let mut snap = HashMap::new();
        for frame in &self.env_stack {
            for (k, v) in frame {
                snap.insert(k.clone(), v.clone());
            }
        }
        snap
    }
}
